File: cpp/server/ad_generator.cpp
#include "ad_generator.h"
#include <functional>
#include <random>
#include <sstream>
#include <iomanip>

AdsList AdGenerator::generateAds(const Context& context, int version) {
    AdsList ads_list;
    ads_list.set_version(version);
    
    // Generate 5-10 ads based on context
    std::hash<std::string> hasher;
    size_t seed = hasher(context.query() + context.asin_id() + std::to_string(version));
    std::mt19937 gen(seed);
    std::uniform_int_distribution<> ad_count_dist(5, 10);
    
    int num_ads = ad_count_dist(gen);
    
    for (int i = 0; i < num_ads; i++) {
        ads::Ad* ad = ads_list.add_ads();
        
        // Generate asin_id based on context and index
        std::stringstream asin_stream;
        asin_stream << "B" << std::setfill('0') << std::setw(6) 
                   << (hasher(context.asin_id() + std::to_string(i)) % 1000000);
        ad->set_asin_id(asin_stream.str());
        
        // Generate ad_id
        ad->set_ad_id(generateAdId(ad->asin_id(), i));
        
        // Calculate score based on context and version
        double score = calculateScore(context.query(), context.asin_id(), 
                                    context.understanding(), version);
        
        // Add some variation per ad
        std::uniform_real_distribution<> score_variation(-0.1, 0.1);
        score += score_variation(gen);
        
        // Clamp score to [0.0, 1.0]
        score = std::max(0.0, std::min(1.0, score));
        ad->set_score(score);
    }
    
    return ads_list;
}

double AdGenerator::calculateScore(const std::string& query, const std::string& asin_id, 
                                 const std::string& understanding, int version) {
    std::hash<std::string> hasher;
    
    // Base score from query and asin_id
    double base_score = static_cast<double>(hasher(query + asin_id) % 1000) / 1000.0;
    
    // Understanding boost (when understanding is provided)
    double understanding_boost = 0.0;
    if (!understanding.empty()) {
        understanding_boost = static_cast<double>(hasher(understanding) % 200) / 1000.0; // 0-0.2 boost
    }
    
    // Version refinement (progressive improvement)
    double version_multiplier = 0.7 + (version * 0.1); // 0.8, 0.9, 1.0 for versions 1, 2, 3
    
    double final_score = (base_score + understanding_boost) * version_multiplier;
    
    // Ensure score is in valid range
    return std::max(0.0, std::min(1.0, final_score));
}

std::string AdGenerator::generateAdId(const std::string& asin_id, int index) {
    std::hash<std::string> hasher;
    size_t hash_value = hasher(asin_id + std::to_string(index));
    
    std::stringstream ad_id_stream;
    ad_id_stream << "AD" << std::setfill('0') << std::setw(8) << (hash_value % 100000000);
    
    return ad_id_stream.str();
}

File: java/src/main/java/com/example/ads/server/AdGenerator.java
package com.example.ads.server;

import ads.Ads;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * Mock ad generation algorithm that creates realistic ads based on Context information.
 * Implements progressive refinement logic across versions 1, 2, and 3.
 */
public class AdGenerator {
    private static final Random random = new Random();
    
    // Mock product categories for realistic ad generation
    private static final String[] PRODUCT_CATEGORIES = {
        "Electronics", "Home & Kitchen", "Books", "Clothing", "Sports", 
        "Beauty", "Automotive", "Tools", "Toys", "Health"
    };
    
    /**
     * Generate a list of mock ads based on the provided context and version.
     * 
     * @param context The search context containing query, asin_id, and understanding
     * @param version The version number (1, 2, or 3) for progressive refinement
     * @return AdsList containing 5-10 mock ads with realistic data
     */
    public Ads.AdsList generateAds(Ads.Context context, int version) {
        List<Ads.Ad> ads = new ArrayList<>();
        
        // Generate 5-10 ads based on context
        int adCount = 5 + random.nextInt(6); // 5-10 ads
        
        for (int i = 0; i < adCount; i++) {
            Ads.Ad ad = generateSingleAd(context, version, i);
            ads.add(ad);
        }
        
        return Ads.AdsList.newBuilder()
                .addAllAds(ads)
                .setVersion(version)
                .build();
    }
    
    /**
     * Generate a single ad with progressive refinement based on version.
     */
    private Ads.Ad generateSingleAd(Ads.Context context, int version, int index) {
        // Base scoring using hash of query + asin_id for deterministic results
        double baseScore = generateBaseScore(context.getQuery(), context.getAsinId(), index);
        
        // Apply version-based refinement
        double refinedScore = applyVersionRefinement(baseScore, context, version);
        
        // Apply understanding boost for versions 2 and 3
        if (version >= 2 && !context.getUnderstanding().isEmpty()) {
            refinedScore = applyUnderstandingBoost(refinedScore, context.getUnderstanding());
        }
        
        // Generate realistic asin_id and ad_id
        String adAsinId = generateAdAsinId(context, index);
        String adId = generateAdId(context, version, index);
        
        return Ads.Ad.newBuilder()
                .setAsinId(adAsinId)
                .setAdId(adId)
                .setScore(Math.min(1.0, Math.max(0.0, refinedScore))) // Clamp to [0.0, 1.0]
                .build();
    }
    
    /**
     * Generate base score using deterministic hash of context information.
     */
    private double generateBaseScore(String query, String asinId, int index) {
        // Create a deterministic hash for consistent results
        int hash = (query + asinId + index).hashCode();
        
        // Convert hash to a score between 0.3 and 0.8 (reasonable base range)
        return 0.3 + (Math.abs(hash) % 1000) / 2000.0;
    }
    
    /**
     * Apply version-based refinement to improve scores progressively.
     */
    private double applyVersionRefinement(double baseScore, Ads.Context context, int version) {
        switch (version) {
            case 1:
                // Version 1: Basic scoring with some randomness
                return baseScore + (random.nextGaussian() * 0.05);
                
            case 2:
                // Version 2: Improved scoring with context consideration
                double improvement = calculateContextRelevance(context) * 0.1;
                return baseScore + improvement + (random.nextGaussian() * 0.03);
                
            case 3:
                // Version 3: Best scoring with full context and understanding
                double maxImprovement = calculateContextRelevance(context) * 0.15;
                if (!context.getUnderstanding().isEmpty()) {
                    maxImprovement += 0.05; // Additional boost for understanding
                }
                return baseScore + maxImprovement + (random.nextGaussian() * 0.02);
                
            default:
                return baseScore;
        }
    }
    
    /**
     * Calculate relevance score based on context information.
     */
    private double calculateContextRelevance(Ads.Context context) {
        double relevance = 0.0;
        
        // Query relevance (longer queries generally indicate more specific intent)
        if (!context.getQuery().isEmpty()) {
            relevance += Math.min(0.3, context.getQuery().length() / 50.0);
        }
        
        // ASIN relevance (having a specific product context is valuable)
        if (!context.getAsinId().isEmpty()) {
            relevance += 0.2;
        }
        
        return relevance;
    }
    
    /**
     * Apply understanding boost for refined queries.
     */
    private double applyUnderstandingBoost(double score, String understanding) {
        // Understanding provides additional context for better targeting
        double boost = Math.min(0.1, understanding.length() / 100.0);
        return score + boost;
    }
    
    /**
     * Generate realistic ASIN ID for the ad.
     */
    private String generateAdAsinId(Ads.Context context, int index) {
        // Use context ASIN as base, or generate related ASINs
        if (!context.getAsinId().isEmpty() && random.nextDouble() < 0.3) {
            // 30% chance to use the same ASIN (exact match)
            return context.getAsinId();
        } else {
            // Generate related ASIN
            String category = PRODUCT_CATEGORIES[Math.abs((context.getQuery() + index).hashCode()) % PRODUCT_CATEGORIES.length];
            return "B" + String.format("%06d", Math.abs((category + index).hashCode()) % 1000000);
        }
    }
    
    /**
     * Generate unique ad ID.
     */
    private String generateAdId(Ads.Context context, int version, int index) {
        // Generate unique ad ID based on context, version, and index
        int hash = (context.getQuery() + context.getAsinId() + version + index).hashCode();
        return "AD" + String.format("%08d", Math.abs(hash) % 100000000);
    }
}